actors/market/src/deal.rs

```diff
@@ -136,12 +136,12 @@ pub struct ClientDealProposal {
 
 #[derive(Clone, Debug, PartialEq, Eq, Copy, Serialize_tuple, Deserialize_tuple)]
 pub struct DealState {
+    // 0 if not yet included in proven sector (0 is also a valid sector number)
+    pub sector_number: SectorNumber,
```

Comment:
What are the implications of changing the serialisation here? Going from `[i64,i64,i64,u64]` to `[u64,i64,i64,i64]` (all just really an `[int,int,int,int]`) with a different mapping of every element. Is this on chain and do these get versioned like specs-actors?

Reply:
slowly catching up as I work through the codebase and figure out how this (kind of works), noting that I'm way behind on my understanding of actors since nv16!

but I still haven't figured out how we deal with these layout changes, do migrations happen in here somewhere, or elsewhere?

Reply:
migration doesn't happen in actors but  the clients(node). for lotus it happen in go state type and https://github.com/filecoin-project/go-state-types/blob/master/builtin/v9/migration/market.go is the last time we migrated deal states as an example.

added more [in this slack thread too.](https://filecoinproject.slack.com/archives/CP50PPW2X/p1704940522070889)

Reply:
Ah, ok, thanks @jennijuju, migrations go into https://github.com/filecoin-project/go-state-types/tree/master/builtin/v13 and so far this isn't represented there.

Next question: the normal process of changing layouts - do they get proposed in here and then implemented in go-state-types later as both the new type and the migration? And what does testing look like that ensures they are match between here and there at the right epoch?

Reply:
I think the first pass on migration for DDO is added here already https://github.com/filecoin-project/go-state-types/pull/236 (will answer the follow up q in slack as its less directly related to this PR!

Reply:
OK, learning more, and this case is particularly interesting because state v12 is storing `DealState` as `[int,int,int,int]` (all unsigned, i.e. no negint possibility even though some of these are `int64`) and the new `DealState` is also `[int,int,int,int]` and the migration shuffles the first 3 across one, drops the last and introduces a new one.

So a migration check that just looks at encoding forms is going to miss any problems because objects could go either way, but their values are going to be jumbled up and meaningless. So a migration probably needs a manual test here to make sure that this change ends up being implemented properly.

Reply:
kk, it's all in here already, and you can see the uint shuffling going on, take one uint off the back and put a uint on the front, 🤞 : https://github.com/filecoin-project/go-state-types/pull/236/files#diff-6354b13bed89e3f3394fb6e8432bf0284ec03f038623eeda03a7b7260846c8bb